### ISSUE TITLE
Verification Counterexamples Are Missing Inputs

### ISSUE THREAD

**Issue by BLuedtke** [kieler/semantics#3]
The counterexample trace generated by KIELER does not specify the required inputs at each tick. For the following nuXmv trace, the following incorrect counterexample is returned:
```
;
;
;
;
;
```
The correct trace should be:
```
a = true;
a = false;
a = true;
a = false;
;
```

# nuXmv Trace
```
nuXmv > go
nuXmv > check_property -P Error_never_occurs
-- invariant !error  is false
-- as demonstrated by the following execution sequence
Trace Description: AG alpha Counterexample 
Trace Type: Counterexample 
  -> State: 1.1 <-
    a = TRUE
    _GO = TRUE
    _pg8 = FALSE
    _pb = FALSE
    _pg39 = FALSE
    _pc = FALSE
    _pg16 = FALSE
    _pg37 = FALSE
    _pg13 = FALSE
    _pg35 = FALSE
    _pg32 = FALSE
    _ppre_a = FALSE
    _pg51 = FALSE
    _preg_a = FALSE
    _perror = FALSE
    _preg__region0_System_a_pre = FALSE
    _ppre__region0_System_a_pre = FALSE
    _pg3 = FALSE
    _pg26 = FALSE
    _pg47 = FALSE
    _pregion0_System_a_pre = FALSE
    _pg43 = FALSE
    _pg21 = FALSE
    c = FALSE
    b = FALSE
    _g18_e2 = TRUE
    _g10_e1 = TRUE
    _cg34 = FALSE
    _cg31 = FALSE
    _cg30 = TRUE
    _cg22 = TRUE
    _cg17 = TRUE
    _cg19 = FALSE
    _cg14 = TRUE
    _cg9 = TRUE
    _cg11 = TRUE
    _cg4 = FALSE
    _cg24 = FALSE
    _cg2 = TRUE
    _g52 = FALSE
    _g51 = TRUE
    _g50 = TRUE
    _g48 = FALSE
    _g47b = TRUE
    _g47 = TRUE
    _g46 = TRUE
    _g44 = FALSE
    _g43 = TRUE
    _g42 = TRUE
    _g40 = FALSE
    _g39 = FALSE
    _g38 = FALSE
    _g37 = TRUE
    _g36b = FALSE
    _g36 = FALSE
    _g35 = FALSE
    _g34 = TRUE
    _g33 = FALSE
    _g32 = FALSE
    _g31 = TRUE
    _g30 = TRUE
    _g29 = TRUE
    _g27 = FALSE
    _g26 = FALSE
    _g25 = FALSE
    _g24 = FALSE
    _g23b = FALSE
    _g23 = FALSE
    _g22 = FALSE
    _g21 = FALSE
    _g20 = FALSE
    _g19 = FALSE
    _g18 = FALSE
    _g17 = FALSE
    _g16 = FALSE
    _g15 = FALSE
    _g14 = FALSE
    _g13 = FALSE
    _g12 = FALSE
    _g11 = FALSE
    _g10 = FALSE
    _g9 = FALSE
    _g8 = FALSE
    _g7 = FALSE
    _g6 = FALSE
    _g5 = FALSE
    _g4b = FALSE
    _g4 = FALSE
    _g3 = TRUE
    _g2 = TRUE
    _g1 = TRUE
    _g0 = TRUE
    _pre_a = FALSE
    _pre_a1 = FALSE
    _pre_a0 = FALSE
    _reg_a = TRUE
    _reg_a3 = TRUE
    _reg_a2 = FALSE
    _reg_a1 = FALSE
    _reg_a0 = FALSE
    _pre__region0_System_a_pre = FALSE
    _pre__region0_System_a_pre1 = FALSE
    _pre__region0_System_a_pre0 = FALSE
    _reg__region0_System_a_pre = FALSE
    _reg__region0_System_a_pre3 = FALSE
    _reg__region0_System_a_pre2 = FALSE
    _reg__region0_System_a_pre1 = FALSE
    _reg__region0_System_a_pre0 = FALSE
    _region0_System_a_pre = FALSE
    _region0_System_a_pre7 = TRUE
    _region0_System_a_pre6 = FALSE
    _region0_System_a_pre5 = TRUE
    _region0_System_a_pre4 = FALSE
    _region0_System_a_pre3 = TRUE
    _region0_System_a_pre2 = FALSE
    _region0_System_a_pre1 = FALSE
    _region0_System_a_pre0 = FALSE
    error = FALSE
    error3 = TRUE
    error2 = FALSE
    error1 = FALSE
    error0 = FALSE
    c3 = TRUE
    c2 = FALSE
    c1 = FALSE
    c0 = FALSE
    b3 = TRUE
    b2 = FALSE
    b1 = FALSE
    b0 = FALSE
  -> State: 1.2 <-
    a = FALSE
    _GO = FALSE
    _pg37 = TRUE
    _pg51 = TRUE
    _preg_a = TRUE
    _pg3 = TRUE
    _pg47 = TRUE
    _pg43 = TRUE
    _cg30 = FALSE
    _cg22 = FALSE
    _cg17 = FALSE
    _cg14 = FALSE
    _cg9 = FALSE
    _cg11 = FALSE
    _cg24 = TRUE
    _cg2 = FALSE
    _g52 = TRUE
    _g50 = FALSE
    _g48 = TRUE
    _g46 = FALSE
    _g44 = TRUE
    _g42 = FALSE
    _g38 = TRUE
    _g30 = FALSE
    _g29 = FALSE
    _g25 = TRUE
    _g24 = TRUE
    _g4b = TRUE
    _g4 = TRUE
    _g2 = FALSE
    _g1 = FALSE
    _g0 = FALSE
    _pre_a = TRUE
    _pre_a1 = TRUE
    _reg_a = FALSE
    _reg_a3 = FALSE
    _reg_a2 = TRUE
    _reg_a0 = TRUE
    _reg__region0_System_a_pre = TRUE
    _reg__region0_System_a_pre3 = TRUE
    _region0_System_a_pre = TRUE
    _region0_System_a_pre6 = TRUE
    _region0_System_a_pre4 = TRUE
  -> State: 1.3 <-
    a = TRUE
    _ppre_a = TRUE
    _preg_a = FALSE
    _preg__region0_System_a_pre = TRUE
    _pregion0_System_a_pre = TRUE
    _cg30 = TRUE
    _cg22 = TRUE
    _cg17 = TRUE
    _cg19 = TRUE
    _cg14 = TRUE
    _cg9 = TRUE
    _cg11 = TRUE
    _cg4 = TRUE
    _cg24 = FALSE
    _cg2 = TRUE
    _g25 = FALSE
    _g24 = FALSE
    _g16 = TRUE
    _g15 = TRUE
    _g8 = TRUE
    _g7 = TRUE
    _g6 = TRUE
    _g5 = TRUE
    _g3 = FALSE
    _pre_a = FALSE
    _pre_a1 = FALSE
    _pre_a0 = TRUE
    _reg_a = TRUE
    _reg_a3 = TRUE
    _reg_a2 = FALSE
    _reg_a0 = FALSE
    _pre__region0_System_a_pre = TRUE
    _pre__region0_System_a_pre1 = TRUE
    _reg__region0_System_a_pre2 = TRUE
    _reg__region0_System_a_pre0 = TRUE
    _region0_System_a_pre4 = FALSE
    _region0_System_a_pre0 = TRUE
  -> State: 1.4 <-
    a = FALSE
    _pg8 = TRUE
    _pg16 = TRUE
    _ppre_a = FALSE
    _preg_a = TRUE
    _ppre__region0_System_a_pre = TRUE
    _pg3 = FALSE
    c = TRUE
    _g18_e2 = FALSE
    _g10_e1 = FALSE
    _cg34 = TRUE
    _cg30 = FALSE
    _cg22 = FALSE
    _cg17 = FALSE
    _cg14 = FALSE
    _cg9 = FALSE
    _cg11 = FALSE
    _cg24 = TRUE
    _cg2 = FALSE
    _g37 = FALSE
    _g35 = TRUE
    _g21 = TRUE
    _g20 = TRUE
    _g19 = TRUE
    _g17 = TRUE
    _g16 = FALSE
    _g15 = FALSE
    _g11 = TRUE
    _g9 = TRUE
    _g7 = FALSE
    _g6 = FALSE
    _g5 = FALSE
    _g4b = FALSE
    _g4 = FALSE
    _pre_a = TRUE
    _pre_a1 = TRUE
    _pre_a0 = FALSE
    _reg_a = FALSE
    _reg_a3 = FALSE
    _reg_a2 = TRUE
    _reg_a0 = TRUE
    _pre__region0_System_a_pre0 = TRUE
    _reg__region0_System_a_pre = FALSE
    _reg__region0_System_a_pre3 = FALSE
    _region0_System_a_pre = FALSE
    _region0_System_a_pre6 = FALSE
  -> State: 1.5 <-
    _pc = TRUE
    _pg16 = FALSE
    _pg37 = FALSE
    _pg35 = TRUE
    _ppre_a = TRUE
    _preg_a = FALSE
    _preg__region0_System_a_pre = FALSE
    _pregion0_System_a_pre = FALSE
    _pg21 = TRUE
    c = FALSE
    b = TRUE
    _cg34 = FALSE
    _cg31 = TRUE
    _cg19 = FALSE
    _cg11 = TRUE
    _cg4 = FALSE
    _cg24 = FALSE
    _g38 = FALSE
    _g36b = TRUE
    _g36 = TRUE
    _g34 = FALSE
    _g31 = FALSE
    _g22 = TRUE
    _g20 = FALSE
    _g19 = FALSE
    _g17 = FALSE
    _g13 = TRUE
    _g12 = TRUE
    _g8 = FALSE
    _pre_a = FALSE
    _pre_a1 = FALSE
    _pre_a0 = TRUE
    _reg_a2 = FALSE
    _reg_a0 = FALSE
    _pre__region0_System_a_pre = FALSE
    _pre__region0_System_a_pre1 = FALSE
    _reg__region0_System_a_pre2 = FALSE
    _reg__region0_System_a_pre0 = FALSE
    _region0_System_a_pre0 = FALSE
    error = TRUE
    c0 = TRUE
nuXmv > quit
```

I will be investigating why exactly this occurs. If I find a solution, there'll be a pull request in the future.

From what I've found so far, Kieler correctly picks up the variable assignments when parsing the counterexample. However, come trace generation, it interprets all these variable assignment expressions to be internal variables. Internal vars are not needed for the counterexample output. Seems like there is something about the variable store that is going wrong. I'll take a closer look.

**Comment by BLuedtke** [comment]
In Class `CompilationContext` (package `de.cau.cs.kieler.kicool.compilation`), the calling of processors to achieve e.g. verification with nuXmv is performed. The processors are called one after the other (see the loops in method @ [L225](https://github.com/kieler/semantics/blob/master/plugins/de.cau.cs.kieler.kicool/src/de/cau/cs/kieler/kicool/compilation/CompilationContext.xtend#L225)). They return the (compilation-) environment with the changes they made (`environmentPrime = processor.compileEntry(environmentPrime)`). The changes to the environment made by the first processor are thus available to the second processor, and so on. Meanwhile, there is also a member "startEnvironment" in the CompilationContext class. This captures the environment _before_ the processors are run (with some exceptions for initialization). Changes to the environment made by processors are usually not saved to this startEnvironment.
Judging by the name "startEnvironment", this is the expected behavior: Said environment should preserve the state of the environment when we started compilation.

Now, when starting verification processors for nuXmv verification, we start with a roughly initialized, but rather sparse environment based on startEnvironment. One of the early processors causes the addition of a "VariableStore" to the working environment, which is passed on to the next processor, and so on. As mentioned earlier, this does not change the startEnvironment, just the local environment that is passed on.
This continues until we arrive at the last processor. This processor is called is "RunNuxmvProcessor", which is essentially just a small addition to "RunSmvProcessor". Here, the task is to both run the nuXmv process, and also interpret its outputs. This concerns the generation of counterexample kTraces: For analyzing the counterexample, the processor relies on the variableStore of the environment to decide whether a variable needs to be put into the trace or not (see method getKTrace in class VerificationPropertyCounterexample). Input and output variables do need to be added, internal vars do not need to be.

The problem is that RunSmvProcessor is searching for the variableStore in the `startEnvironment` when analyzing the results, and not in `environment`. See [line 178](https://github.com/kieler/semantics/blob/master/plugins/de.cau.cs.kieler.verification/src/de/cau/cs/kieler/verification/processors/nuxmv/RunSmvProcessor.xtend#L178) in Class `RunSmvProcessor` in package `de.cau.cs.kieler.verification.processors.nuxmv` in the verification plugin. The startEnvironment does not contain the variable store, an empty store is thus returned. As the variableStore is empty, all variables are interpreted as internal, and none are added to the KTrace.

The fix should be rather simple: Just use `environment` instead of `startEnvironment` - the former is passed to the processor anyway, so it should never be null. If it does not contain a variableStore for whatever reason, we have the same situation as now, the store will be added to the context, but it'll be empty. Some testing is required nonetheless. I'll open a pull request (which will also include this recap) once the fix is ready.
